src/mon/OSDMonitor.cc

```diff
@@ -4586,6 +4693,90 @@ void OSDMonitor::tick()
   if (do_propose ||
       !pending_inc.new_pg_temp.empty())  // also propose if we adjusted pg_temp
     propose_pending();
+
+  if (ceph_using_tcmalloc() && pcm != nullptr) {
+    _adjust_cache_ratios();
+    pcm->balance();
+    pcm->tune_memory();
+    _set_new_cache_sizes();
+  }
+}
+
+void OSDMonitor::_set_new_cache_sizes()
+{
+  uint64_t cache_size = 0;
+  int64_t inc_alloc = 0;
+  int64_t full_alloc = 0;
+  int64_t kv_alloc = 0;
+
+  if (pcm != nullptr && rocksdb_binned_kv_cache != nullptr) {
+    cache_size = pcm->get_tuned_mem();
+    inc_alloc = inc_cache->get_committed_size();
+    full_alloc = full_cache->get_committed_size();
+    kv_alloc = rocksdb_binned_kv_cache->get_committed_size();
+  }
+
+  uint64_t max_inc_cache = static_cast<uint64_t>(
+    inc_alloc / mon_memory_target / inc_cache->get_avg_bytes_per_osdmap());
+  // set full cache size to be equal to inc cache size
+  uint64_t max_full_cache = max_inc_cache;
+
+  dout(1) << __func__ << " cache_size:" << cache_size
+          << " inc_alloc: " << inc_alloc
+          << " full_alloc: " << full_alloc
+          << " kv_alloc: " << kv_alloc
+          << dendl;
+  dout(1) << " avg inc bytes per osdmap: "
+          << inc_cache->get_avg_bytes_per_osdmap()
+          << " max inc cache: " << max_inc_cache
+          << dendl;
+  inc_osd_cache.set_size(max_inc_cache);
+  full_osd_cache.set_size(max_full_cache);
+}
+
+uint64_t OSDMonitor::_get_used_inc_cache_bytes()
+{
+  dout(1) << __func__ << " Calculate inc used bytes from oldest to newest epoch" << dendl;
+  uint64_t total_bytes = 0;
+  uint64_t features = mon->get_quorum_con_features();
+  epoch_t first = get_first_committed();
+  epoch_t latest = osdmap.get_epoch();
+
+  for (epoch_t e = latest; e >= first && e > 0; e--) {
```

Comment:
calling get_version() or get_version_full() here will read off of disk if the {epoch, features} version is not found in the corresponding cache. It'll be much more efficient to keep track of the size in the incremental and full osdmap caches themselves, rather than recalculating each tick().

Reply:
Sure, I have changed the logic to track the used bytes within the incremental and full size caches.